connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -1063,6 +1076,112 @@ public int generation() {
+            restartRequests = new ArrayList<>(pendingRestartRequests.values());
+            pendingRestartRequests.clear();
+        }
+        for (RestartRequest restartRequest : restartRequests) {
+            try {
+                doRestartConnectorAndTasks(restartRequest);
+            } catch (Exception e) {
+                log.warn("Unexpected error while trying to process " + restartRequest + ", the restart request will be skipped.", e);
+            }
+        }
+    }
+
+    protected synchronized void doRestartConnectorAndTasks(RestartRequest request) {
+        final String connectorName = request.connectorName();
+        Optional<RestartPlan> maybePlan = buildRestartPlanFor(request);
+        if (!maybePlan.isPresent()) {
+            log.debug("Skipping restart of connector '{}' since no status is available: {}", connectorName, request);
+            return;
+        }
+        RestartPlan plan = maybePlan.get();
+        log.info("Executing {}", plan);
+
+        // If requested, stop the connector and any tasks, marking each as restarting
+        final ExtendedAssignment currentAssignments = assignment;
+        final Collection<ConnectorTaskId> assignedIdsToRestart = plan.taskIdsToRestart()
+                                                                     .stream()
+                                                                     .filter(taskId -> currentAssignments.tasks().contains(taskId))
+                                                                     .collect(Collectors.toList());
+        final boolean restartConnector = plan.restartConnector() && currentAssignments.connectors().contains(connectorName);
+        final boolean restartTasks = !assignedIdsToRestart.isEmpty();
+        if (restartConnector) {
+            worker.stopAndAwaitConnector(connectorName);
+            recordRestarting(connectorName);
+        }
+        if (restartTasks) {
+            // Stop the tasks and mark as restarting
+            worker.stopAndAwaitTasks(assignedIdsToRestart);
+            assignedIdsToRestart.forEach(this::recordRestarting);
+        }
+
+        // Now restart the connector and tasks
+        if (restartConnector) {
+            startConnector(connectorName, (error, targetState) -> {
+                if (error == null) {
+                    log.info("Connector '{}' successfully restarted", connectorName);
+                } else {
+                    log.error("Failed to restart connector '" + connectorName + "'", error);
+                }
+            });
+        }
+        if (restartTasks) {
+            log.debug("Restarting {} of {} tasks for {}", plan.restartTaskCount(), plan.totalTaskCount(), request);
+            plan.taskIdsToRestart().forEach(this::startTask);
+            log.debug("Restarted {} of {} tasks for {} as requested", plan.restartTaskCount(), plan.totalTaskCount(), request);
+        }
```

Comment:
Can you confirm that the methods called in this block of logic handle failures? For example, the `worker.stopAndAwaitConnector(...)` method waits only up to 5 seconds, and we have no visibility into whether the connector was actually stopped in that time or we timed out. If we timed out, should we restart the connector, or should we abort this method and try again later? If we retry again later, it seems like it's okay to try stopping an already stopped connector, but it'd be good to double check that.

Same questions about stopping the tasks.

We may need to modify the `stopAndAwait...(...)` methods to add a callback that will allow us to better track what was actually stopped. And where this method is called we may need to handle the case where this restart was not completed and re-enqueue it for restart.

Reply:
@rhauch I did confirm that both stop and await for task and connector does connector.cancel() and task.cancel() if it timed out. I added try/catch in start connector/tasks https://github.com/apache/kafka/pull/10822/commits/5b5debe2d42a8d67e08cdb4998ef267fb87f9385

However, I am not sure about re-enqueue as it can lead to infinite tries, and adding retries may complicate things so need some guidance. Could you please review to see if it looks a bit better now?

Reply:
@rhauch we can add a retryCount in the RestartRequest and start with a default of 5 and decrement it every time we re-enqueue. If the tick() happens every second then we can exhaust the retries pretty fast so we may need to add backoff with a delay before processing the request. Adding retry logic to this PR can make the PR complex, I would vote for adding it in a separate PR.

Reply:
> Adding retry logic to this PR can make the PR complex, I would vote for adding it in a separate PR.

I would agree. We're not sure whether we need this behavior yet, so I'm +1 for keeping it simple in this PR and adding it in the future only if we discover a need for it.